In [1]:
import numpy as np
import pandas as pd

/tmp/ipykernel_30873/1662815981.py:2: DeprecationWarning: 
Pyarrow will become a required dependency of pandas in the next major release of pandas (pandas 3.0),
(to allow more performant data types, such as the Arrow string type, and better interoperability with other libraries)
but was not found to be installed on your system.
If this would cause problems for you,
please provide us feedback at https://github.com/pandas-dev/pandas/issues/54466
        
  import pandas as pd


In [2]:
# import the dataset
df = pd.read_csv("sample_0_ptrac.csv");
df

Type          X          Y          Z         U         V         W  \
0    1000   0.000000   0.000000   0.000000  0.508477  0.473278  0.719347   
1    3000  26.153789  24.343315  37.000000  0.508477  0.473278  0.719347   
2    4000  28.567736  26.590158  40.415032  0.508477  0.473278  0.719347   
3    5000  28.567736  26.590158  40.415032  0.508477  0.473278  0.719347   
4    2000  28.567736  26.590158  40.415032  0.642856 -0.725381  0.246084   
..    ...        ...        ...        ...       ...       ...       ...   
427  5000  42.260018  11.133571  45.659893  0.956422  0.291984  0.001599   
428  2000  42.259198  11.136909  45.658301  0.320801  0.453103  0.831736   
429  5000  42.259198  11.136909  45.658302 -0.043768 -0.699589  0.713204   
430  2000  42.258623  11.139260  45.657183  0.096900  0.258149  0.961233   
431  5000  42.258623  11.139261  45.657186 -0.333001  0.020514  0.942703   

        Energy  Weight  Cell  Material      Time  
0    14.000000     1.0  30.0       6.0  0.000000  
1    14.000000     1.0  20.0       1.0  1.004945  
2    14.000000     1.0  20.0       1.0  1.097699  
3    14.000000     1.0  20.0       1.0  1.097699  
4     2.429137     1.0  20.0       1.0  1.097699  
..         ...     ...   ...       ...       ...  
427   0.000982     1.0  20.0       1.0  1.168771  
428   0.001087     1.0  20.0       1.0  1.168757  
429   0.000993     1.0  20.0       1.0  1.168757  
430   0.001121     1.0  20.0       1.0  1.168748  
431   0.000729     1.0  20.0       1.0  1.168748  

[432 rows x 12 columns]

In [4]:
df.columns

Index(['Type', 'X', 'Y', 'Z', 'U', 'V', 'W', 'Energy', 'Weight', 'Cell',
       'Material', 'Time'],
      dtype='object')

In [5]:
df['Type'].unique()

array([1000, 3000, 4000, 5000, 2000])

In [6]:
df['Type'].value_counts()

Type
5000    204
2000    203
4000     23
1000      1
3000      1
Name: count, dtype: int64

Starts with one particle.  
203 new particles were created.   
204 (all particles) were destroyed.  

In [7]:
is_increasing = all(df['Time'].iloc[i] > df['Time'].iloc[i-1] for i in range(1, len(df)))

if is_increasing:
    print("The time column is strictly increasing.")
else:
    print("The time column is not strictly increasing.")

The time column is not strictly increasing.


In [8]:
# Check if the time column is strictly decreasing
is_decreasing = all(df['Time'].iloc[i] < df['Time'].iloc[i-1] for i in range(1, len(df)))

if is_decreasing:
    print("The time column is strictly decreasing.")
else:
    print("The time column is not strictly decreasing.")

The time column is not strictly decreasing.


In [9]:
# Initialize variables
datasets = []
current_dataset = []

# Iterate over the DataFrame
for index, row in df.iterrows():
    event_type = row['Type']
    
    # Check if the current row is the start of a new dataset
    if event_type == 2000:
        if current_dataset:  # Append the current dataset if it's not empty
            datasets.append(pd.DataFrame(current_dataset))
            current_dataset = []  # Reset the current dataset
        
    # Append the current row to the current dataset
    current_dataset.append([row['Type'], row['X'], row['Y'], row['Z'], row["Time"]])
    
# Append the last dataset
if current_dataset:
    datasets.append(pd.DataFrame(current_dataset))

# Print the number of datasets created
print("Number of datasets:", len(datasets))


Number of datasets: 204


In [10]:
# uncomment to print each dataset
# for idx, dataset in enumerate(datasets, start=1):
#     print(f"Dataset {idx}:")
#     print(dataset)
#     print("\n") 

In [11]:
# printing the datasets where the particle dies as soon as it is created
for idx, dataset in enumerate(datasets, start=1):
    if len(dataset) == 2:
        print(f"Dataset {idx}:")
        # print(dataset) # uncomment to print the datasets
        print("\n")
        
        # Calculate coordinate differences and time difference
        x_diff = dataset.iloc[1][1] - dataset.iloc[0][1]
        y_diff = dataset.iloc[1][2] - dataset.iloc[0][2]
        z_diff = dataset.iloc[1][3] - dataset.iloc[0][3]
        time_diff = dataset.iloc[1][4] - dataset.iloc[0][4]
        
        # Print coordinate differences and time difference
        print("Coordinate Differences:")
        print(f"  X-coordinate Difference: {x_diff:.6f}")
        print(f"  Y-coordinate Difference: {y_diff:.6f}")
        print(f"  Z-coordinate Difference: {z_diff:.6f}")
        print(f"  Time Difference: {time_diff:.6f}")
        print("\n")

Dataset 3:


Coordinate Differences:
  X-coordinate Difference: -0.000864
  Y-coordinate Difference: -0.001222
  Z-coordinate Difference: 0.000349
  Time Difference: 0.000042


Dataset 4:


Coordinate Differences:
  X-coordinate Difference: 0.000008
  Y-coordinate Difference: -0.000036
  Z-coordinate Difference: -0.000017
  Time Difference: 0.000002


Dataset 5:


Coordinate Differences:
  X-coordinate Difference: -0.000048
  Y-coordinate Difference: 0.000032
  Z-coordinate Difference: 0.000092
  Time Difference: 0.000004


Dataset 6:


Coordinate Differences:
  X-coordinate Difference: -0.000050
  Y-coordinate Difference: 0.000022
  Z-coordinate Difference: -0.000168
  Time Difference: 0.000006


Dataset 7:


Coordinate Differences:
  X-coordinate Difference: 0.000004
  Y-coordinate Difference: 0.000007
  Z-coordinate Difference: 0.000039
  Time Difference: 0.000003


Dataset 8:


Coordinate Differences:
  X-coordinate Difference: 0.000000
  Y-coordinate Difference: -0.000000
  Z-coor

In [13]:
# printing the datasets that have more than 2 rows
for idx, dataset in enumerate(datasets, start=1):
    if len(dataset) > 2:
        print(f"Dataset {idx}")
        print(dataset)
        print("\n")

Dataset 1
        0          1          2          3         4
0  1000.0   0.000000   0.000000   0.000000  0.000000
1  3000.0  26.153789  24.343315  37.000000  1.004945
2  4000.0  28.567736  26.590158  40.415032  1.097699
3  5000.0  28.567736  26.590158  40.415032  1.097699


Dataset 2
         0          1          2          3         4
0   2000.0  28.567736  26.590158  40.415032  1.097699
1   4000.0  42.257788  11.142679  45.655559  1.168734
2   4000.0  44.316991  10.528963  45.542828  1.175911
3   4000.0  45.787397  11.055095  47.935630  1.185442
4   4000.0  38.756944  17.174070  48.410751  1.216572
5   4000.0  36.259014  19.504108  49.087581  1.228188
6   4000.0  37.246132  17.249322  51.136169  1.238870
7   4000.0  42.338193  19.694256  45.077795  1.266500
8   4000.0  42.682331   8.300546  38.348026  1.310654
9   4000.0  45.101685  10.208199  38.599819  1.320966
10  4000.0  44.436272   8.738580  38.901607  1.326440
11  4000.0  43.635166   8.642429  38.132947  1.330157
12  4000.0 

So most of the particles are terminated as soon as they are created almost at the same coordinates.  
We have five cases where the particle traverses before it is terminated.

In [14]:
df['Time'].value_counts()

Time
1.097699    3
1.170030    3
1.310665    3
1.363131    3
1.336365    3
           ..
1.176632    1
1.176658    1
1.176658    1
1.176704    1
1.168748    1
Name: count, Length: 400, dtype: int64

In [15]:
duplicate_time_rows  = df[df.duplicated(subset=['Time'], keep=False)][['Type', 'Time']].sort_values(by="Time")
print(duplicate_time_rows)
print("Length", len(duplicate_time_rows))

     Type      Time
2    4000  1.097699
3    5000  1.097699
4    2000  1.097699
5    4000  1.168734
263  2000  1.168734
398  2000  1.169261
399  5000  1.169261
350  2000  1.170030
349  5000  1.170030
348  4000  1.170030
268  5000  1.171053
267  2000  1.171053
6    4000  1.175911
117  2000  1.175911
71   2000  1.185442
7    4000  1.185442
114  5000  1.185491
113  2000  1.185491
8    4000  1.216572
69   2000  1.216572
9    4000  1.228188
53   2000  1.228188
170  4000  1.236052
203  2000  1.236052
215  2000  1.236154
216  5000  1.236154
47   2000  1.238870
10   4000  1.238870
11   4000  1.266500
43   2000  1.266500
171  4000  1.299784
199  2000  1.299784
12   4000  1.310654
34   2000  1.310654
41   2000  1.310665
40   5000  1.310665
39   4000  1.310665
13   4000  1.320966
32   2000  1.320966
14   4000  1.326440
28   2000  1.326440
26   2000  1.333164
16   4000  1.333164
193  2000  1.333828
172  4000  1.333828
24   2000  1.336171
17   4000  1.336171
20   2000  1.336365
19   5000  1.336365


There are cases where multiple events occured at the same time.

In [16]:
# Group the DataFrame by the "Time" column
grouped_df = df.groupby('Time')

# Sort the groups by time
sorted_groups = {time: group.sort_values(by='Time') for time, group in grouped_df}

# Map indexes to sorted time values
mapped_indexes = {time: group.index.tolist() for time, group in sorted_groups.items()}
mapped_indexes

{0.0: [0],
 1.0049448587218393: [1],
 1.0976994253366426: [2, 3, 4],
 1.1687340998415632: [5, 263],
 1.168747685966123: [430],
 1.168747815951192: [431],
 1.1687570296081768: [428],
 1.168757107723021: [429],
 1.1687703030390613: [426],
 1.1687707935151634: [427],
 1.1688055442340173: [422],
 1.1688064060053664: [424],
 1.1688083150024395: [423],
 1.1688146765738672: [425],
 1.1688672978940218: [420],
 1.1688676160498457: [421],
 1.1689169722120916: [418],
 1.1689171692562308: [419],
 1.168946662163262: [416],
 1.1689501291424396: [417],
 1.169010928007467: [414],
 1.169011255563224: [415],
 1.1690287805635735: [412],
 1.1690288503765789: [413],
 1.1690791330524632: [408],
 1.169079256438433: [409],
 1.1690895428705872: [410],
 1.169094813173769: [411],
 1.1691367052496944: [406],
 1.16913684069743: [407],
 1.1691949117438618: [404],
 1.169195114455635: [405],
 1.1692016997258516: [402],
 1.1692021824504266: [403],
 1.169236939982676: [400],
 1.1692373193839949: [401],
 1.1692613720006

In [9]:
# checking for the rows with same coordinates
coor_duplicates = df[(df.duplicated(subset=['X', 'Y', 'Z'], keep=False)) & (df['Type'].isin([2000, 5000]))]

grouped_coor = coor_duplicates.groupby(['X', 'Y', 'Z']).filter(lambda x: len(x) >= 2)

for index, group in grouped_coor.groupby(['X', 'Y', 'Z']):
    print(f"\nGroup: {index}")
    print(group[['Type', 'X', 'Y', 'Z']])

print(f"\nNum of duplicate coordinate rows with Type 2000 or 5000: {len(grouped_coor)}")



Group: (28.567735930838523, 26.59015800232582, 40.41503211341633)
   Type          X          Y          Z
3  5000  28.567736  26.590158  40.415032
4  2000  28.567736  26.590158  40.415032

Group: (42.14000349791876, 10.9446615134013, 45.82571592041534)
     Type          X          Y          Z
349  5000  42.140003  10.944662  45.825716
350  2000  42.140003  10.944662  45.825716

Group: (42.19055063407367, 10.869224941920216, 45.91765197843671)
     Type          X          Y          Z
267  2000  42.190551  10.869225  45.917652
268  5000  42.190551  10.869225  45.917652

Group: (42.26562579136401, 11.0073553906202, 45.71038311335081)
     Type          X          Y          Z
398  2000  42.265626  11.007355  45.710383
399  5000  42.265626  11.007355  45.710383

Group: (42.68465589390481, 8.298929511715865, 38.34922537713809)
    Type          X        Y          Z
40  5000  42.684656  8.29893  38.349225
41  2000  42.684656  8.29893  38.349225

Group: (43.982647562969625, 8.806252449

There are 20 cases where the particle is spawned and terminated in the same coordinate.